x/staking/abci.go

```diff
@@ -50,7 +50,12 @@ func EndBlocker(ctx sdk.Context, k keeper.Keeper) []abci.ValidatorUpdate {
 			k.DeleteByKey(ctx, iterator.Key())
 		}
 
-		return k.BlockValidatorUpdates(ctx)
+		// defer TODO should update epochNumber after epoch finish
+		// This won't affect slashing module since slashing Endblocker run before staking module
+
+		return k.EpochValidatorUpdates(ctx)
```

Comment:
Wouldn't it be okay to always do ValidatorUpdates at the end of every block?  Because if the staking messages haven't taken effect, then unless there was a slash, the validator set just wouldn't update because there were no changes?

Reply:
I see, for CreateValidator we don't buffer the message so that users can delegate to inactive validators that's created before end of epoch.

Reply:
For the CreateValidator, we can do the creation of the validator immediately, but buffer the self-delegation